### PR TITLE
Attempt to write metric tests with inmemoryexp

### PR DESCRIPTION
**Pull request by cijothomas**
Attempting to write Metrics test to ensure SDK is doing the aggregation as expected. ~This is full of "hacks" to unblock.! Need to spend some time to understand the inner workings to fix this - If anyone can help, really appreciate it!~ Thanks @TommyCpp  who [suggested the fix](https://github.com/open-telemetry/opentelemetry-rust/pull/1382#discussion_r1398257009) for flush being stuck!

Such tests are very critical to make sure the aggregation is correct, especially considering we are considering big changes like the ones in this PR - https://github.com/open-telemetry/opentelemetry-rust/pull/1379

### CODE REVIEW

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 1]
```diff
@@ -57,3 +57,86 @@ pub use meter_provider::*;
...
+
+        // TODO: Unable to get this to work, as
+        // flush is taking forever.
+        // println!("flush started");
+        // meter_provider.force_flush().unwrap();
+        // println!("flush finished");
+        // The sleep is temporary workaround to
+        // unblock progress on writing tests like
+        // this.
+        sleep(Duration::from_millis(100)).await;
```
Comment (cijothomas) [inline opentelemetry-sdk/src/metrics/mod.rs:105]: its risky to check-in this as-is, as this won't be reliable. So need to understand why flush/shutdown is not working inside unittests, and fix it..
@jtescher @TommyCpp Either of you can help given you have much more knowledge of tokio runtime/metric sdk internals!

Comment (TommyCpp) [reply opentelemetry-sdk/src/metrics/mod.rs:105]: Tokio test by default uses single thread runtime, which can be tricky. See [here](https://docs.rs/tokio/latest/tokio/attr.test.html)

Maybe try annotate the test to use multi thread runtime

Comment (cijothomas) [reply opentelemetry-sdk/src/metrics/mod.rs:105]: That did fix the Flush-not-progressing issue! Thank you!
I'll try to tweak the tests to make it reliable now.

**Review by hdost — APPROVED** [review]
Looks like a winner to me 🤘 one step closer to stability.

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 2]
```diff
@@ -57,3 +57,103 @@ pub use meter_provider::*;
...
+            .init();
+        counter.add(1, &[KeyValue::new("key1", "value1")]);
+        counter.add(1, &[KeyValue::new("key1", "value1")]);
+        counter.add(1, &[KeyValue::new("key1", "value1")]);
+        counter.add(1, &[KeyValue::new("key1", "value1")]);
+        counter.add(1, &[KeyValue::new("key1", "value1")]);
+
+        counter.add(1, &[KeyValue::new("key1", "value2")]);
+        counter.add(1, &[KeyValue::new("key1", "value2")]);
+        counter.add(1, &[KeyValue::new("key1", "value2")]);
```
Comment (lalitb) [inline opentelemetry-sdk/src/metrics/mod.rs:96]: nit - attributes can be defined once, and reused - something like (not tested):
```cpp
let key_value1 = KeyValue::new("key1", "value1");
let key_value2 = KeyValue::new("key1", "value2");

for _ in 0..5 {
    counter.add(1, &[key_value1]); 

for _ in 0..3 {
    counter.add(1, &[key_value2]);
}
```

**Review by lalitb — APPROVED** [review]
LGTM. Nicely done :)

**Comment by cijothomas** [comment]
This is far from ideal, but merging as is, give this works, and can be used to quickly validate that aggregations/merges are done correctly, especially when working on PRs like https://github.com/open-telemetry/opentelemetry-rust/pull/1379

Expecting to improve such tests in follow ups. Thanks everyone for helping review!

**Pull request merged**